theano/tensor/opt.py

```diff
@@ -320,7 +320,50 @@ def inplace_elemwise_optimizer(fgraph):
             raised_warning = not verbose
 
             for candidate_output in candidate_outputs:
-                for candidate_input in candidate_inputs:
+
+                # If the output of the node can be established as an update
+                # output of the fgraph, visit the candidate_inputs in an order
+                # that will improve the chances of making the node operate
+                # inplace on the input it's meant to update
+                candidate_out_var = node.outputs[candidate_output]
+                sorted_candidate_inputs = candidate_inputs
+                if fgraph.update_mapping:
+
+                    update_outs = [fgraph.outputs[i]
+                                   for i in fgraph.update_mapping]
+
+                    if candidate_out_var in update_outs:
+
+                        # The candidate output is an update. Sort the
+                        # variables in candidate_inputs in the following order:
+                        # - Vars corresponding to the actual updated input
+                        # - Vars computed inplace on the updates input
```

Comment:
Those 2 lines are very short. It took me a few minutes to remember the second case. Can you explain more why this is needed? To help the next reader of that method.